src/algebra/cubic_discriminant.lean

```diff
@@ -0,0 +1,359 @@
+
+## Tags
+
+cubic, discriminant, polynomial, root
+-/
+
+noncomputable theory
+
+/-- The structure representing a cubic polynomial. -/
+@[ext] structure cubic (R : Type*) := (a b c d : R)
+
+instance (R : Type*) [has_zero R] : inhabited (cubic R) := ⟨⟨0, 0, 0, 0⟩⟩
+
+namespace cubic
+
+open cubic polynomial
+
+/-! ## Coefficients -/
+
+section coeff
+
+variables {R S : Type*} [semiring R] [semiring S] {P Q : cubic R} (φ : R →+* S)
+
+/-- Convert a cubic polynomial to a polynomial. -/
+def to_poly (P : cubic R) : polynomial R := C P.d + C P.c * X + C P.b * X ^ 2 + C P.a * X ^ 3
+
+/-- Map a cubic polynomial across a ring homomorphism. -/
+def map (P : cubic R) : cubic S := ⟨φ P.a, φ P.b, φ P.c, φ P.d⟩
+
+lemma map_to_poly : (map φ P).to_poly = polynomial.map φ P.to_poly :=
+by simp only [map, to_poly, map_C, map_X, polynomial.map_add, polynomial.map_mul,
+              polynomial.map_pow]
+
+lemma coeffs :
+  P.to_poly.coeff 3 = P.a ∧ P.to_poly.coeff 2 = P.b ∧ P.to_poly.coeff 1 = P.c
+    ∧ P.to_poly.coeff 0 = P.d :=
+by { simp only [to_poly, coeff_add, coeff_C, coeff_C_mul, coeff_X, coeff_C_mul_X], norm_num }
+
+@[simp] theorem coeff_three : P.to_poly.coeff 3 = P.a := coeffs.1
+
+@[simp] theorem coeff_two : P.to_poly.coeff 2 = P.b := coeffs.2.1
+
+@[simp] theorem coeff_one : P.to_poly.coeff 1 = P.c := coeffs.2.2.1
+
+@[simp] theorem coeff_zero : P.to_poly.coeff 0 = P.d := coeffs.2.2.2
+
+theorem a_of_eq (h : P.to_poly = Q.to_poly) : P.a = Q.a := by rw [← coeff_three, h, coeff_three]
+
+theorem b_of_eq (h : P.to_poly = Q.to_poly) : P.b = Q.b := by rw [← coeff_two, h, coeff_two]
+
+theorem c_of_eq (h : P.to_poly = Q.to_poly) : P.c = Q.c := by rw [← coeff_one, h, coeff_one]
+
+theorem d_of_eq (h : P.to_poly = Q.to_poly) : P.d = Q.d := by rw [← coeff_zero, h, coeff_zero]
+
+lemma to_poly_injective (P Q : cubic R) : P.to_poly = Q.to_poly ↔ P = Q :=
+⟨λ h, cubic.ext _ _ (a_of_eq h) (b_of_eq h) (c_of_eq h) (d_of_eq h), congr_arg _⟩
+
+lemma of_a_eq_zero (ha : P.a = 0) : P.to_poly = C P.d + C P.c * X + C P.b * X ^ 2 :=
+by rw [to_poly, C_eq_zero.mpr ha, zero_mul, add_zero]
+
+lemma of_b_eq_zero (ha : P.a = 0) (hb : P.b = 0) : P.to_poly = C P.d + C P.c * X :=
+by rw [to_poly, C_eq_zero.mpr ha, C_eq_zero.mpr hb, zero_mul, zero_mul, add_zero, add_zero]
+
+lemma of_c_eq_zero (ha : P.a = 0) (hb : P.b = 0) (hc : P.c = 0) : P.to_poly = C P.d :=
+by rw [to_poly, C_eq_zero.mpr ha, C_eq_zero.mpr hb, C_eq_zero.mpr hc, zero_mul, zero_mul, zero_mul,
+       add_zero, add_zero, add_zero]
+
+lemma of_zero (ha : P.a = 0) (hb : P.b = 0) (hc : P.c = 0) (hd : P.d = 0) : P.to_poly = 0 :=
+by rw [to_poly, C_eq_zero.mpr ha, C_eq_zero.mpr hb, C_eq_zero.mpr hc, C_eq_zero.mpr hd, zero_mul,
+       zero_mul, zero_mul, add_zero, add_zero, add_zero]
+
+end coeff
+
+/-! ## Degree -/
+
+section degree
+
+variables {R : Type*} [semiring R] {P : cubic R}
+
+lemma degree_of_constant_lt : (C P.d).degree < 1 := lt_of_le_of_lt degree_C_le dec_trivial
```

Comment:
I believe that all of the lemmas in this section should really be one-liners. I'm not too familiar with the polynomial library, but I suggest that you go into `data/polynomnial/degree/definitions.lean` or `data/polynomial/degree/lemmas.lean`, start with a variant like
```lean
lemma degree_add_le' {P Q : polynomial R} {n : with_bot ℕ} (hP : degree P ≤ n) (hQ : degree Q ≤ n) :
  degree (P + Q) ≤ n :=
(degree_add_le P Q).trans (max_le hP hQ)
```
and then prove general versions of the lemmas you are proving here, so that you can derive all of these lemmas as one-liners.

With the lemma I mentioned above, you can also give a shorter direct proof of `degree_le`:
```lean
theorem degree_le : P.to_poly.degree ≤ 3 :=
begin
  refine (degree_add_le' (degree_add_le' (degree_add_le' _ _) _) _),
  { exact degree_C_le.trans dec_trivial },
  { exact (degree_C_mul_X_le _).trans dec_trivial },
  all_goals { exact (degree_C_mul_X_pow_le _ _).trans dec_trivial },
end
```